src/Service/RegenerateProductUrl.php

```diff
@@ -204,18 +177,28 @@ private function log(string $message): void
     }
 
     /**
-     * Add product ulrs
-     *
      * @param array $urls
-     * @param bool $last
+     * @param bool  $last
      *
      * @return int
-     * @throws UrlAlreadyExistsException
      */
     private function replaceUrls(array &$urls, bool $last = false): int
     {
         $this->log(sprintf('replaceUrls%s batch: %d', $last ? ' last' : '', count($urls)));
-        $this->urlPersist->replace($urls);
+
+        foreach ($urls as $url) {
+            try {
+                $this->urlPersist->replace([$url]);
+            } catch (\Exception $e) {
+                $this->log(
+                    sprintf(
+                        $e->getMessage() . ' Entity id: %d Request path: %s',
+                        $url->getEntityId(),
+                        $url->getRequestPath()
+                    )
+                );
+            }
+        }
```

Comment:
@Gerke-W @peterjaap I think we could provide a loop to process each url incase the runner wants to have a log trace with error products. In this case looping the batch then process each url one would decrease performance a bit.
Here my performance testing

With **foreach on batch**:
![image](https://github.com/user-attachments/assets/4aa3016a-5057-40ab-8e84-84747a7d6da6)

Without looping inside the batch:
![image](https://github.com/user-attachments/assets/f86e4fde-1499-4dd8-ad7d-4b47fd419428)

Reply:
LGTM! PR it up!